ext/debug/ctf/README.rst

```diff
@@ -0,0 +1,204 @@
+and graphical (TraceCompass) variants already exist.
+
+CTF should look familiar to C programmers but adds stronger typing.
+See `CTF - A  Flexible, High-performance Binary Trace Format
+<http://diamon.org/ctf/>`_.
+
+Every system has application-specific events to trace out.  Historically,
+that has implied:
+
+1. Determining the application-specific payload,
+2. Choosing suitable serialization-format,
+3. Writing the on-target serialization code,
+4. Deciding on and writing the IO transport mechanics,
+5. Writing the PC-side deserializer/parser,
+6. Writing custom ad-hoc tools for filtering and presentation.
+
+CTF allows us to formally describe #1 and #2, which enables common
+infrastructure for #5 and #6.  This leaves #3 serialization code and #4
+IO mechanics up to a custom implementation.
+
+This CTF debug module aims at providing a common #1 and #2 for Zephyr
+("middle"), while providing a lean & generic interface for IO ("bottom").
+Currently, only one CTF bottom-layer exists, POSIX ``fwrite``, but many others
+are possible:
+
+- Async UART
+- Async DMA
+- Sync GPIO
+- ... and many more.
+
+In fact, IO varies greatly from system to system.  Therefore, it is
+instructive to create a taxonomy for IO types when we must ensure the
+interface between CTF-middle and CTF-bottom is generic and efficient
+enough to model these. See the *IO taxonomy* section below.
+
+
+A generic interface
+-------------------
+
+In CTF, an event is serialized to a packet containing one or more fields
+and as seen from *IO taxonomy* section below, a bottom layer:
+
+- may perform action at transaction-start (e.g. mutex-lock),
+- may process each field in some way (e.g. sync-push emit, concat, enqueue to thread-bound FIFO),
+- may perform action at transaction-stop (e.g. mutex-release, emit of concat buffer).
+
+This is possible when the bottom-layer implements the following macros:
+
+- ``CTF_BOTTOM_LOCK``:   No-op or how to lock the IO transaction.
+- ``CTF_BOTTOM_UNLOCK``: No-op or how to release the IO transaction.
+- ``CTF_BOTTOM_FIELDS``: Var-args of fields. May process each field with ``MAP``.
+- ``CTF_BOTTOM_TIMESTAMPED_INTERNALLY``: Tells where timestamping is done.
+
+These macros along with inline functions of the middle-layer can yield
+very low-overhead tracing infrastructure.
+
+
+CTF middle-layer example
+------------------------
+
+The CTF_EVENT macro will serialize each argument to a field::
+
+  /* Example for illustration */
+  static inline void ctf_middle_foo(u32_t thread_id, ctf_bounded_string_t name)
+  {
+    CTF_EVENT(
+      CTF_LITERAL(u8_t, 42),
+      thread_id,
+      name,
+      "hello, I was emitted from function: ",
+      __func__  /* __func__ is standard since C99 */
+    );
+  }
+
+How to serialize and emit fields as well as handling alignment, can be done
+internally and statically at compile-time in the bottom-layer.
+
+
+How to activate?
+----------------
+
+Make sure ``CONFIG_CTF=y`` and ``CONFIG_CTF_BOTTOM_POSIX=y`` is set.
```

Comment:
you can set ```CONFIG_CTF_BOTTOM_POSIX``` as default in the native_posix board if CONFIG_CTF is set.
See for example https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/posix/native_posix/Kconfig.defconfig#L33
That way it is enough to set CONFIG_CTF, and for the native_posix board it would automatically go thru BOTTOM_POSIX